usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -489,6 +519,467 @@ static CK_RV cleanse_attribute(TEMPLATE *template,
+     * ep11_data. This should always be the case, because we just
+     * created the pkey with the current MK. */
+    vp_offset = pkey_attr->ulValueLen - MK_VP_LENGTH;
+    if (memcmp(&ep11_data->pkey_mk_vp, pkey_attr->pValue + vp_offset,
+               MK_VP_LENGTH) != 0) {
+        TRACE_ERROR("vp of this pkey does not match with the one in ep11_data (should not occur)\n");
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* Now update the key obj. If it's a token obj, it will be also updated
+     * in the repository. */
+    ret = pkey_update_and_save(tokdata, key_obj, pkey_attr);
+    if (ret != CKR_OK) {
+        TRACE_ERROR("pkey_update_and_save failed with rc=0x%lx\n", ret);
+        goto done;
+    }
+
+    ret = CKR_OK;
+
+done:
+
+    return ret;
+}
+
+/**
+ * Returns true if the session is ok for creating protected keys, false
+ * otherwise. The session must be read/write and not public, because the
+ * protected key is a private/secret key attribute.
+ */
+static CK_BBOOL ep11tok_pkey_session_read_write(SESSION *session)
+{
+    if (session->session_info.flags & CKF_RW_SESSION)
+        return CK_TRUE;
+
+    return CK_FALSE;
+}
+
+/**
+ * Checks if the preconditions for using the related protected key of
+ * the given secure key object are met. The caller of this routine must
+ * have a READ_LOCK on the key object.
+ *
+ * @return true if conditions met
+ *         false otherwise
+ */
+CK_BBOOL ep11tok_pkey_usage_ok(STDLL_TokData_t *tokdata, SESSION *session,
+                               OBJECT *key_obj, CK_MECHANISM *mech)
+{
+    ep11_private_data_t *ep11_data = tokdata->private_data;
+    CK_ATTRIBUTE *opaque_attr = NULL;
+    CK_BBOOL bval = CK_FALSE;
+    CK_RV ret;
+
+    /* Check if CPACF supports the operation implied by this key and mech */
+    if (!pkey_op_supported_by_cpacf(mech))
+        goto done;
+
+    /* Check config option */
+    switch (ep11_data->pkey_mode) {
+    case PKEY_MODE_DISABLED:
+        goto done;
+        break;
+    case PKEY_MODE_DEFAULT:
+    case PKEY_MODE_ENABLE4NONEXTR:
+        /* Use existing pkeys, re-create invalid pkeys, and also create new
+         * pkeys for keys that do not already have one. */
+        if (object_is_extractable(key_obj) ||
+            !object_is_pkey_extractable(key_obj) ||
+            object_is_attr_bound(key_obj) ||
+            !ep11tok_pkey_session_read_write(session) ||
+            !ep11_data->pkey_wrap_supported) {
+            goto done;
+        }
+        if (template_attribute_get_non_empty(key_obj->template,
+                                             CKA_IBM_OPAQUE_PKEY,
+                                             &opaque_attr) != CKR_OK ||
+            !ep11tok_pkey_is_valid(tokdata, key_obj)) {
+            /* this key has either no pkey attr, or it is not valid,
+             * try to create one */
+            ret = ep11tok_pkey_update(tokdata, key_obj);
+            if (ret != CKR_OK) {
```

Comment:
Because pkey_update_and_save (called within ep11tok_pkey_update) unlocks and relocks the object, it might get concurrently updated by other threads at this place. 

Since you just return from this function after returning from ep11tok_pkey_update it should be save for this function, but how about the callers?

Reply:
see above.